barman/config.py

```diff
@@ -1314,6 +1320,228 @@ def _validate_with_keys(config_items, allowed_keys, section):
+
+        Generates configuration change objects starting from a dictionary with
+        the same fields.
+
+        """
+        if set(obj.keys()) == set(cls._fields):
+            return cls(**obj)
+        if set(obj.keys()) == set(ConfigChange._fields):
+            return ConfigChange(**obj)
+        raise ValueError("Malformed configuration change serialization: %r" % obj)
+
+
+class ConfigChangesQueue:
+    """
+    Wraps the management of the config changes queue.
+
+    Once instantiated the queue can be accessed using the `queue` property.
+    """
+
+    def __init__(self, queue_file):
+        self.queue_file = queue_file
+        self._queue = None
+        self.open()
+
+    @staticmethod
+    def read_file(path) -> List[ConfigChangeSet]:
+        """Reads a json file containing a list of configuration changes."""
+        try:
+            with open(path, "r") as queue_file:
+                # Read the queue if exists
+                return json.load(queue_file, object_hook=ConfigChangeSet.from_dict)
+        except FileNotFoundError:
+            return []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    @property
+    def queue(self):
+        if self._queue is None:
+            self.open()
+        else:
+            return self._queue
+
+    def open(self):
+        self._queue = self.read_file(self.queue_file)
+
+    def close(self):
+        with open(self.queue_file + ".tmp", "x") as queue_file:
+            # Dump the configuration change list into the queue file
+            json.dump(self._queue, queue_file, cls=ConfigChangeSetEncoder, indent=2)
+
+        # Juggle with the queue files to ensure consistency of
+        # the queue even if Shelver is interrupted abruptly
+        old_file_name = self.queue_file + ".old"
+        try:
+            os.rename(self.queue_file, old_file_name)
+        except FileNotFoundError:
+            old_file_name = None
+        os.rename(self.queue_file + ".tmp", self.queue_file)
+        if old_file_name:
+            os.remove(old_file_name)
+        self._queue = None
+
+
+class ConfigChangesProcessor:
+    """
+    The class is responsible for processing the config changes to
+    apply to the barman config
+    """
+
+    def __init__(self, config):
+        self.config = config
+        self.applied_changes = []
+
+    def receive_config_changes(self, changes):
+        """
+        Process all the configuration changes
+        """
+        # Get all the available configuration change files in order
+        changes_list = []
+        cfg_json = json.loads(changes)
+        for section in cfg_json:
+            set = ConfigChangeSet(section)
+            for json_cng in cfg_json[section]:
+                chg = ConfigChange(
+                    json_cng,
+                    cfg_json[section][json_cng],
+                    self.config.config._config.get_config_source(section, json_cng),
```

Comment:
side comment: `self.config.config._config` looks so weird 😆

Reply:
Heh, me neither!

Reply:
>Maybe ~/.barman.conf could be a good place, as suggested by Giulio.

I did various tests and in the end the simplest solution is: if is default dump everything in ~/.barman.conf

We just need to document it. 
IF we cannot write in the config file (ie config file in /etc/ with just read rights) we want to die horribly until the rights are updated... opinions on this @barthisrael @mikewallace1979 ?

Reply:
I guess that makes sense, yes.
I mean, files under `/etc` have higher priority than `~/.barman.conf`, so we need to let the user know if we fail to update an `/etc` file as writing to `~/.barman.conf` would not take effect in that case because of the setting being defined in these two places.